### PR TITLE
Add TimeoutError exception.

### PR DESCRIPTION
**Pull request by danicampora**
Added to alias it to socket.timeout. It's important to be able to tell the difference between a `socket.error` and `socket.timeout`, therefore we need separate exceptions. TimeoutError is subclassed from OSError (as per CPython docs) and inside the socket module it is aliased to socket.timeout. An extra commit is also included to show how this would apply to the CC3200.

### CODE REVIEW

**Comment by pfalcon** [comment]
From https://docs.python.org/3/library/exceptions.html#TimeoutError : 

> Raised when a system function timed out at the system level. Corresponds to errno ETIMEDOUT.

In other words, OSError contains the same amount of information anyway. Yes, TimeoutError is slightly easier to access, but what percentage of people will actually use it? Likely small, and those can use OSError and check args[0]. So, all in all, is it worth to add it, given that try to optimize for size a lot of other things?

**Comment by danicampora** [comment]
> Yes, TimeoutError is slightly easier to access, but what percentage of people will actually use it?

I would say a lot, specially when working with sockets. It is very common to see code like this:

```python
    try:
        r = sock.recv(length)
    except socket.timeout:
        return b''
```

it can of course be rewritten to:

```python
    try:
        r = sock.recv(length)
    except OSError as e:
        if e.args[0] == ETIMEDOUT:
            pass
        else:
            raise
```

But the natural way is to use the first form, and having it improves usability a lot.
In the case of `socket.error`, yes, it is deprecated because it's an alias of OSError, but still the common thing to do is to check for `socket.error` when dealing with internet stuff. I think that this patch will make using the usocket module a lot more straightforward.

I am the first one that tries to save bytes in every corner, but I've also realized that usability is equally important.

**Comment by dpgeorge** [comment]
> But the natural way is to use the first form, and having it improves usability.

Not to mention that catching `socket.timeout` is the existing and (I assume) recommended way of coding such things.

I see this not as a discussion about adding functionality (which as @pfalcon points out is already there with OSError(ETIMEDOUT)), but as adding compatibility with scripts written for (C)Python.

Regarding:
```
try:
    r = sock.recv(length)
except OSError as e:
    if e.args[0] == ETIMEDOUT:
        return b''
    else:
        raise
```
this won't work in CPython because `socket.timeout` has empty args.  Therefore, using OSError as uPy's socket.timeout would make it hard to write code that ran correctly under both CPython and uPy.

**Comment by danicampora** [comment]
> this won't work in CPython because socket.timeout has empty args.

True, good point.

**Comment by danicampora** [comment]
Is it OK to merge this?

**Comment by pfalcon** [comment]
No probs on my side, and I'd say it was mergeable after @dpgeorge's comment.

**Comment by danicampora** [comment]
Thanks, just wanted to double check :-)

**Comment by dpgeorge** [comment]
Yes.

File: py/mpconfig.h [hunk 1]
```diff
@@ -462,6 +462,11 @@ typedef double mp_float_t;
 #define MICROPY_PY_BUILTINS_RANGE_ATTRS (1)
 #endif
 
+// Wether to support timeout exceptions (like socket.timeout)
```
Comment (dpgeorge) [inline py/mpconfig.h]: Spelling mistake in "Wether".

Comment (danicampora) [inline py/mpconfig.h]: Go it!

File: py/obj.h [hunk 2]
```diff
@@ -426,6 +426,9 @@ extern const mp_obj_type_t mp_type_MemoryError;
 extern const mp_obj_type_t mp_type_NameError;
 extern const mp_obj_type_t mp_type_NotImplementedError;
 extern const mp_obj_type_t mp_type_OSError;
+#if MICROPY_PY_BUILTINS_TIMEOUTERROR
```
Comment (dpgeorge) [inline py/obj.h]: No need for this #if guard (it's not generally used for optional header declarations to keep things cleaner).

Comment (danicampora) [inline py/obj.h]: OK!

**Comment by danicampora** [comment]
Corrected stuff pointed out by @dpgeorge and merged in https://github.com/micropython/micropython/commit/077812b2abe3f5e5325194f9694dad7eb38186dd

**Pull request closed without merging**